### PR TITLE
service: nfc: Implement NFC IUser service

### PR DESCRIPTION
**Pull request by german77**
NFP, Miifare and NFC on the core use the same code. The only difference are the validation checks and specialized calls. Stubs `SendCommandByPassThrough` while I find documentation on what these commands are.

Fixes `Ultra Kaiju Monster Rancher` nfc detection.

### CODE REVIEW

File: src/core/hle/service/nfc/nfc_user.cpp [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    for (auto& device : devices) {
+        if (nfp_devices.size() >= max_allowed_devices) {
+            continue;
+        }
+        if (device->GetCurrentState() != NFP::DeviceState::Unavailable) {
+            nfp_devices.push_back(device->GetHandle());
+        }
+    }
+
+    if (nfp_devices.size() == 0) {
```
Comment (lioncash) [inline src/core/hle/service/nfc/nfc_user.cpp:131]:
```suggestion
    if (nfp_devices.empty()) {
```

File: src/core/hle/service/nfc/nfc_device.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <array>
+#include <atomic>
```
Comment (lioncash) [inline src/core/hle/service/nfc/nfc_device.cpp:5]: Doesn't seem like this or `<array>` are used anywehere

File: src/core/hle/service/nfc/nfc_device.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <array>
```
Comment (lioncash) [inline src/core/hle/service/nfc/nfc_device.h:6]: Ditto

File: src/core/hle/service/nfc/nfc_device.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+#include <vector>
+
+#include "common/common_funcs.h"
```
Comment (lioncash) [inline src/core/hle/service/nfc/nfc_device.h:9]:
```suggestion
#include "common/common_types.h"
```

File: src/core/hle/service/nfc/nfc_user.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <array>
+#include <optional>
```
Comment (lioncash) [inline src/core/hle/service/nfc/nfc_user.h:7]: Missed this in the previous review, but `<memory>` needs to be included.

File: src/core/hle/service/nfp/nfp_user.h [hunk 6]
```diff
@@ -3,6 +3,9 @@
 
 #pragma once
 
+#include <array>
+#include <optional>
```
Comment (lioncash) [inline src/core/hle/service/nfp/nfp_user.h:7]: ditto

File: src/core/hle/service/nfc/nfc_device.cpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+    Core::HID::ControllerUpdateCallback engine_callback{
+        .on_change = [this](Core::HID::ControllerTriggerType type) { NpadUpdate(type); },
+        .is_npad_service = false,
+    };
+    is_controller_set = true;
+    callback_key = npad_device->SetCallback(engine_callback);
+}
+
+NfcDevice::~NfcDevice() {
```
Comment (liamwhite) [inline src/core/hle/service/nfc/nfc_device.cpp:34]: Close the events you opened

File: src/core/hle/service/nfc/nfc_user.cpp [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+        {412, &IUser::AttachDeactivateEvent, "AttachDeactivateEvent"},
+        {1000, nullptr, "ReadMifare"},
+        {1001, nullptr, "WriteMifare"},
+        {1300, &IUser::SendCommandByPassThrough, "SendCommandByPassThrough"},
+        {1301, nullptr, "KeepPassThroughSession"},
+        {1302, nullptr, "ReleasePassThroughSession"},
+    };
+    RegisterHandlers(functions);
+
+    availability_change_event = service_context.CreateEvent("IUser:AvailabilityChangeEvent");
```
Comment (liamwhite) [inline src/core/hle/service/nfc/nfc_user.cpp:44]: Close the event in the destructor

**Review by liamwhite — APPROVED** [review]

**Review by bunnei — APPROVED** [review]

**Pull request merged**